packages/mgt-chat/src/components/Chat/Chat.tsx

```diff
@@ -47,6 +72,25 @@ export const Chat = ({ chatId }: IMgtChatProps) => {
       chatClient.offStateChange(setChatState);
     };
   }, [chatClient]);
+
+  const onRenderMessage = (messageProps: MessageProps, defaultOnRender?: MessageRenderer): JSX.Element => {
+    if (isChatMessage(messageProps?.message)) {
+      const content = messageProps.message?.content;
+
+      // Test that the content is supported.
+      if (content && unsupportedContentRegex.test(content)) {
+        messageProps = produce(messageProps, (draft: MessageProps) => {
```

Comment:
Having produce in doing work here during the render of data feels wrong to me.

Can we move this processing into the StatefulClient and perhaps modify the chat message during the processing we're already doing on the messages? Maybe add a containsUnsupportedContent  bool to the message object so that the renderer here is fairly dumb?

Reply:
@gavinbarron I'm struggling with this particular challenge where I need to change the value of the `content` from the provided content string to our `UnupportedContent` component value.

I agree that the best approach with the processing is to do it in the StatefulClient. However, since I wanted to use a react component (ease of updating, customizing etc) I can't use it in the StatefulClient. Therefore, doing something like this in the StatefulClient can't work:

```ts
...
if(this.isUnsupportedContent){
  content = renderToString(<UnsupportedContent />)
}
...
```

Removing the react component all together means, I have to create the unsupported content component string in html and inline CSS.

```ts
...
if(this.isUnsupportedContent){
  content = `<a href="#" target="blank" style="LOTS OF STYLING"><icon up/><p>View this message in Microsoft Teams</p></div>`
}
...
```

On the other hand, processing in the `Chat` component means that I have the props for the message, but updating of the props can't work because props are read-only. I introduced `immer` produce to cater for this here. I also agree that it's an overkill and antipattern.

So I'm left in a limbo here because I can only have one of the two approaches above. What would be the lesser evil or alternative?

Reply:
My thinking was that this should be a blend of the two approaches.  
In StatefulClient graphChatMessageToAcsChatMessage, after emoji replacement but prior to image processing
```
if (hasUnspportedContent(content)) {
  result.currentValue = buildUnspportedContentMessage(graphMessage) // this returns an AcsMessage with an additional property to indicated unsupported content
}
```

In the render override:
```
if (message.hasUnspportedContent) {
  // use unspportedContentMessage
} else {
 // use normal rendering
}
```

This way the supported/unsupported content detection is done when we're already processing the message contents and simply setting some data that is used in the rendering process to make a choice about what elements are used to render a thing.

Reply:
Yes, this is also my thought and current update. The challenge is with `message` being a prop and being the container for the content that we need to replace. I wanted to avoid building this entire message rendering area _again_ while it already exists. it's not exposed to allow setting the message content that's why I'm substituting the content with `renderToString` for it to just render it as a chat message:
![image](https://github.com/microsoftgraph/microsoft-graph-toolkit/assets/8081536/13804743-23d7-4e83-9c49-8d3e39b5d7b7)

I'm exploring another angle with the `MessageRenderer` passed as part of the `onRenderMessage` function override. This function does away with everything in the message area. So, another option is to make the `UnsupportedContent` element render the entire message section with the avatar, dates and redirect CTA.